File: app/src/main/java/sixue/naivereader/data/Book.java
package sixue.naivereader.data;

import android.support.annotation.NonNull;

import com.fasterxml.jackson.annotation.JsonIgnore;

import java.util.ArrayList;
import java.util.List;

import sixue.naivereader.helper.BookHelper;
import sixue.naivereader.helper.LocalTextHelper;
import sixue.naivereader.helper.OnlineHelper;
import sixue.naivereader.helper.PacketHelper;

public class Book {
    private String id;
    private String title;
    private String author;
    private BookKind kind;
    private String localPath;
    private String siteId;
    private String sitePara;
    @JsonIgnore
    private List<Chapter> chapterList;
    private List<Source> sources;
    private int currentChapterIndex;
    private int currentPosition;
    private int wordCount;
    private String coverSavePath;
    private boolean end;
    @JsonIgnore
    private BookHelper bookHelper;

    public Book() {
        id = "";
        title = "";
        author = "";
        siteId = "";
        sitePara = "";
        chapterList = new ArrayList<>();
        sources = new ArrayList<>();
        coverSavePath = "";
        end = false;
        bookHelper = null;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public BookKind getKind() {
        return kind;
    }

    public void setKind(BookKind kind) {
        this.kind = kind;
    }

    public String getLocalPath() {
        return localPath;
    }

    public void setLocalPath(String localPath) {
        this.localPath = localPath;
    }

    public String getSiteId() {
        return siteId;
    }

    public void setSiteId(String siteId) {
        this.siteId = siteId;
    }

    public String getSitePara() {
        return sitePara;
    }

    public void setSitePara(String sitePara) {
        this.sitePara = sitePara;
    }

    public List<Chapter> getChapterList() {
        return chapterList;
    }

    public void setChapterList(List<Chapter> chapterList) {
        this.chapterList = chapterList;
    }

    public List<Source> getSources() {
        return sources;
    }

    public void setSources(List<Source> sources) {
        this.sources = sources;
    }

    public int getCurrentChapterIndex() {
        return currentChapterIndex;
    }

    public void setCurrentChapterIndex(int currentChapterIndex) {
        this.currentChapterIndex = currentChapterIndex;
    }

    public int getCurrentPosition() {
        return currentPosition;
    }

    public void setCurrentPosition(int currentPosition) {
        this.currentPosition = currentPosition;
    }

    public int getWordCount() {
        return wordCount;
    }

    public void setWordCount(int wordCount) {
        this.wordCount = wordCount;
    }

    public String getCoverSavePath() {
        return coverSavePath;
    }

    public void setCoverSavePath(String coverSavePath) {
        this.coverSavePath = coverSavePath;
    }

    public boolean isEnd() {
        return end;
    }

    public void setEnd(boolean end) {
        this.end = end;
    }

    @NonNull
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("{");
        sb.append("id=");
        sb.append(id);
        sb.append(",");
        sb.append("title=");
        sb.append(title);
        sb.append(",");
        sb.append("author=");
        sb.append(author);
        sb.append(",");
        sb.append("kind=");
        sb.append(kind);
        if (localPath != null) {
            sb.append(",");
            sb.append("localPath=");
            sb.append(localPath);
        }
        if (siteId != null) {
            sb.append(",");
            sb.append("siteId=");
            sb.append(siteId);
        }
        if (sitePara != null) {
            sb.append(",");
            sb.append("sitePara=");
            sb.append(sitePara);
        }
        sb.append("}");
        return sb.toString();
    }

    public BookHelper buildHelper() {
        if (bookHelper == null) {
            if (kind == BookKind.LocalText) {
                bookHelper = new LocalTextHelper(this);
            } else if (kind == BookKind.Online) {
                bookHelper = new OnlineHelper(this);
            } else if (kind == BookKind.Packet) {
                bookHelper = new PacketHelper(this);
            } else {
                throw new IllegalArgumentException("Unknown KIND: " + kind.toString());
            }
        }
        return bookHelper;
    }
}

File: app/src/main/java/sixue/naivereader/data/Book.kt
package sixue.naivereader.data

import com.fasterxml.jackson.annotation.JsonIgnore
import sixue.naivereader.helper.*
import java.util.*
import kotlin.collections.ArrayList

data class Book (
    var id: String,
    var title: String,
    var author: String,
    var kind: BookKind,
    var localPath: String? = null,
    var siteId: String? = null,
    var sitePara: String? = null,

    var sources: List<Source> = ArrayList(),
    var currentChapterIndex: Int = -1,
    var currentPosition: Int = -1,
    var wordCount: Int = -1,
    var coverSavePath: String = "",
    var end: Boolean = false,
) {
    @JsonIgnore
    var chapterList: List<Chapter> = ArrayList()
    @JsonIgnore
    var localChapterNodes: List<Int> = ArrayList()
    @JsonIgnore
    var localChapterTitles: List<String> = ArrayList()
    @JsonIgnore
    var localChapterSummaries: List<String> = ArrayList()

    @JsonIgnore
    private lateinit var bookHelper: BookHelper
    override fun toString(): String {
        val sb = StringBuilder()
        sb.append("{")
        sb.append("id=")
        sb.append(id)
        sb.append(",")
        sb.append("title=")
        sb.append(title)
        sb.append(",")
        sb.append("author=")
        sb.append(author)
        sb.append(",")
        sb.append("kind=")
        sb.append(kind)
        if (localPath != null) {
            sb.append(",")
            sb.append("localPath=")
            sb.append(localPath)
        }
        if (siteId != null) {
            sb.append(",")
            sb.append("siteId=")
            sb.append(siteId)
        }
        if (sitePara != null) {
            sb.append(",")
            sb.append("sitePara=")
            sb.append(sitePara)
        }
        sb.append("}")
        return sb.toString()
    }

    @JsonIgnore
    fun isRefreshable(): Boolean {
        return kind === BookKind.Online
    }

    @JsonIgnore
    fun isViewableInBrowser(): Boolean {
        return when {
            kind === BookKind.Online -> true
            kind === BookKind.Archive -> true
            else -> false
        }
    }

    fun buildHelper(): BookHelper {
        if (!this::bookHelper.isInitialized) {
            bookHelper = when (kind) {
                BookKind.LocalText -> {
                    LocalTextHelper(this)
                }
                BookKind.Online -> {
                    OnlineHelper(this)
                }
                BookKind.Packet -> {
                    PacketHelper(this)
                }
                BookKind.Archive -> {
                    ArchiveHelper(this)
                }
            }
        }
        return bookHelper
    }
}